contracts/adapters/OpenseaDownpaymentBuyAdapter.sol

```diff
@@ -0,0 +1,180 @@
+  IWETH public WETH;
+
+  struct Params {
+    // bend params
+    address nftAsset;
+    uint256 nftTokenId;
+    // opensea params
+    address[14] addrs;
+    uint256[18] uints;
+    uint8[8] feeMethodsSidesKindsHowToCalls;
+    bytes calldataBuy;
+    bytes calldataSell;
+    bytes replacementPatternBuy;
+    bytes replacementPatternSell;
+    bytes staticExtradataBuy;
+    bytes staticExtradataSell;
+    uint8[2] vs;
+    bytes32[5] rssMetadata;
+  }
+
+  modifier onlyAaveLendPool() {
+    require(msg.sender == aaveAddressedProvider.getLendingPool(), "Caller must be aave lending pool");
+    _;
+  }
+
+  function initialize(
+    address _bendAddressesProvider,
+    address _aaveAddressesProvider,
+    address _weth,
+    address _openseaExchange
+  ) external initializer {
+    __Pausable_init();
+    __Ownable_init();
+    __ReentrancyGuard_init();
+    bendAddressesProvider = ILendPoolAddressesProvider(_bendAddressesProvider);
+    aaveAddressedProvider = IAaveLendPoolAddressesProvider(_aaveAddressesProvider);
+    WETH = IWETH(_weth);
+    openseaExchange = IOpenseaExchage(_openseaExchange);
+  }
+
+  function executeOperation(
+    address[] calldata assets,
+    uint256[] calldata amounts,
+    uint256[] calldata premiums,
+    address initiator,
+    bytes calldata params
+  ) external override nonReentrant whenNotPaused onlyAaveLendPool returns (bool) {
+    // Decode params
+    Params memory _params = _decodeParams(params);
+
+    // Check other parameters
+    require(assets.length == 1 && amounts.length == 1 && premiums.length == 1, "Multiple assets not supported");
+    require(assets[0] == address(WETH), "Only WETH borrowing allowed");
+
+    // Check order payment token
+    address _buyerpaymentToken = _params.addrs[6];
+    address _sellerpaymentToken = _params.addrs[13];
+    require(_buyerpaymentToken == address(0), "Buyer payment token should be ETH");
+    require(_sellerpaymentToken == address(0), "Seller payment token should be ETH");
+
+    uint256 _flashBorrowedAmount = WETH.balanceOf(address(this));
+    assert(amounts[0] == _flashBorrowedAmount); // Should nerver happened
+
+    // Check if the flash loan can be paid off
+    uint256 _flashLoanDebt = _flashBorrowedAmount + premiums[0];
+    (, , uint256 availableBorrows, , , , ) = _getBendLendPool().getNftCollateralData(_params.nftAsset, address(WETH));
+    require(availableBorrows >= _flashLoanDebt, "Unable to repay the loan");
+
+    // Check payment sufficient
+    address _buyer = initiator;
+    uint256 _sellPrice = _params.uints[13];
+    uint256 _buyerBalance = WETH.balanceOf(_buyer);
+    require(_buyerBalance + _flashBorrowedAmount >= _sellPrice, "Insufficient payment");
+
+    // Transfer buyer WETH for payment
+    uint256 _buyerPayment = _sellPrice - _flashBorrowedAmount;
+    WETH.transferFrom(_buyer, address(this), _buyerPayment);
+    WETH.withdraw(_sellPrice);
+
+    // Do opensea exchange
+    openseaExchange.atomicMatch_{value: _sellPrice}(
```

Comment:
Is OpenSea use same api to deal with CryptoPunks?

Reply:
No, must be NFT with ERC721 standard, WPunks is OK